vcd/resource_vcd_edgegateway.go

```diff
@@ -0,0 +1,238 @@
+			},
+			"default_gateway_network": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "",
+				ForceNew:    true,
+				Description: "External network to be used as default gateway. Its name must be included in 'external_networks'. An empty value will skip the default gateway",
+			},
+			"distributed_routing": &schema.Schema{
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				ForceNew:    true,
+				Description: "If advanced networking enabled, also enable distributed routing",
+			},
+		},
+	}
+}
+
+// Creates a new edge gateway from a resource definition
+func resourceVcdEdgeGatewayCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[TRACE] edge gateway creation initiated")
+
+	vcdClient := meta.(*VCDClient)
+
+	rawExternalNetworks := d.Get("external_networks").([]interface{})
+	var externalNetworks []string
+	for _, en := range rawExternalNetworks {
+		externalNetworks = append(externalNetworks, en.(string))
+	}
+
+	// Making sure the parent entities are available
+	orgName := vcdClient.getOrgName(d)
+	vdcName := vcdClient.getVdcName(d)
+
+	var missing []string
+	if orgName == "" {
+		missing = append(missing, "org")
+	}
+	if vdcName == "" {
+		missing = append(missing, "vdc")
+	}
+	if len(missing) > 0 {
+		return fmt.Errorf("missing properties. %v should be given either in the resource or at provider level", missing)
+	}
+
+	org, vdc, err := vcdClient.GetOrgAndVdc(orgName, vdcName)
+	if err != nil {
+		return err
+	}
+	if (org == govcd.Org{}) || org.Org.HREF == "" || org.Org.ID == "" || org.Org.Name == "" {
+		return fmt.Errorf("no valid Organization named '%s' was found", orgName)
+	}
+	if (vdc == govcd.Vdc{}) || vdc.Vdc.HREF == "" || vdc.Vdc.ID == "" || vdc.Vdc.Name == "" {
+		return fmt.Errorf("no valid VDC named '%s' was found", vdcName)
+	}
+
+	var gwCreation = govcd.EdgeGatewayCreation{
+		ExternalNetworks:          externalNetworks,
+		Name:                      d.Get("name").(string),
+		OrgName:                   orgName,
+		VdcName:                   vdcName,
+		Description:               d.Get("description").(string),
+		BackingConfiguration:      d.Get("configuration").(string),
+		AdvancedNetworkingEnabled: d.Get("advanced").(bool),
+		DefaultGateway:            d.Get("default_gateway_network").(string),
+		DistributedRoutingEnabled: d.Get("distributed_routing").(bool),
+		HAEnabled:                 d.Get("ha_enabled").(bool),
+	}
+
+	// In version 9.7+ the advanced property is true by default
+	if vcdClient.APIVCDMaxVersionIs(">= 32.0") {
+		if !gwCreation.AdvancedNetworkingEnabled {
+			return fmt.Errorf("'advanced' property for vCD 9.7+ must be set to 'true'")
+		}
+	}
+
+	edge, err := govcd.CreateEdgeGateway(vcdClient.VCDClient, gwCreation)
+	if err != nil {
+		log.Printf("[DEBUG] Error creating edge gateway: %#v", err)
+		return fmt.Errorf("error creating edge gateway: %#v", err)
+	}
+
+	d.SetId(edge.EdgeGateway.ID)
```

Comment:
I probably missed the discussion on this. Could you remind me what is the main reason we are setting ID here instead of the name as in other resources?

Reply:
It's for future usage with data sources. Currently, only load balancing entities can find items by ID, but I saw what Dainius did for it, and it can be easily applied to every resource.